Multi Page

paddyo opened the pull request:
Work In Progress: Tidied things up a bit in preparation for printing. Added capability to have a separate page for printing, keeping the main page displaying the card. Next thing is the actual print out.

paddyo committed:
Quick tidy up and simplify

paddyo committed:
remove commited deps

paddyo committed:
Add print window page

paddyo committed:
printing functionality

paddyo committed:
add image precache, bug fixes, tarot preview on print page

paddyo committed:
add setup step at beginning

martin-cs commented on sworker.js:
```diff
@@ -0,0 +1,36 @@
+const get_image_names = () => {
+  const total = 101;
+  const image_names = [];
+  for (let n = 0; n < total; n++) {
```
The images are numbered from `j-1.jpg` upward, not `j-0.jpg`.  However there is a `j-.jpg` so maybe rename that or alternatively start from 1?

martin-cs commented on sketch.js:
```diff
@@ -1,36 +1,110 @@
+const KEY_FREE = 'free';
+
+// FIXME: Warning: naive interpolate.
+const translate = (x) => Math.round((100 / 255) * x);
```
There are a couple of issues with this :
- If I understand the semantics of JavaScript `Math.round` as being round-to-nearest then this can give results in [0,100] which is 101 possible outcomes.
- Because there are 256 inputs and 101 outputs, some outputs will happen for 2 inputs, some for 3 which will give a noticeable probability skew.
- (Interestingly, floating-point representation is not an issue as `100 / 255` is exactly representable as a float.)

Possible fixes:
A. If the random number is >=200, throw it away and get another one.  Otherwise return `(randomNumber % 100) + 1` so you have a number in `[1,100]` where each outcome has equal probability.
B. Alternatively change the API call to ask for a `uint16` and return `(randomNumber % 100) + 1`.  This won't be quite uniformly random but the differences will be 65 vs. 66, not 2 vs. 3.

ashmgdesign committed:
reshuffled blank 'j-.jpg' so real range is 'j-1' to 'j-100'

range should ignore old 'j-.jpg' that was a semi-duplicate of 'j-1,jpg'.
Correct sampling range is 'j-1.jpg' to 'j-100.jpg', never should 'j-.jpg' be used - have renamed it to 'j-old-1.jpg'
